### PR TITLE
Streamline actor shutdown logic

### PR DESCRIPTION
**Pull request by mavam**

### CODE REVIEW

File: libvast/vast/system/terminator.hpp [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+struct terminator_state {
+  std::vector<caf::actor_addr> remaining_actors;
+  caf::response_promise promise;
+  static inline const char* name = "terminator";
+};
+
+template <class Policy>
+caf::behavior terminator(caf::stateful_actor<terminator_state>* self);
+
+/// @relates terminator
```
Comment (lava) [inline libvast/vast/system/terminator.hpp:43]: From an API perspective, it feels a bit weird to me that the caller is supposed to create the terminator and then do nothing with it but pass it into this function.

Maybe we should just have functions `shutdown_sequential()` and `shutdown_generic()` (or `shutdown<Policy>()`) that create the terminator internally?

Also, I'm wondering if it would make sense to return the promise here, so the caller can wait for successful termination directly, without having to use `attach_functor()`.

File: libvast/src/system/node.cpp [hunk 2]
```diff
@@ -196,7 +197,7 @@ status_command(const command::invocation&, caf::actor_system& sys) {
 }
 
 maybe_actor spawn_accountant(node_actor* self, spawn_arguments&) {
-  auto accountant = self->spawn<monitored>(system::accountant);
+  auto accountant = self->spawn<linked>(system::accountant);
```
[the remark this answers is not in the document]
Comment (mavam) [reply libvast/src/system/node.cpp:200]: I thought about this more. It's always possible that an actor receives multiple EXIT messages. What's important is that the processing is idempotent, i.e., the shutdown process is initiated only once. In the `shutdown` helper function, I've implemented this by resetting the EXIT handler so that it only performs a hard kill when another EXIT message with reason `kill` arrives.

File: libvast/src/system/node.cpp [hunk 3]
```diff
@@ -389,13 +383,23 @@ void node_state::init(std::string init_name, path init_dir) {
...
-  tracker = self->spawn<monitored>(system::tracker, name);
-  self->set_down_handler([=](const down_msg& msg) {
-    VAST_IGNORE_UNUSED(msg);
-    VAST_DEBUG(self, "got DOWN from", msg.source);
-    self->quit(msg.reason);
+  tracker = self->spawn<linked>(system::tracker, name);
+  self->set_exit_handler([=](const exit_msg& msg) {
+    VAST_DEBUG(self, "got EXIT from", msg.source);
+    auto xs = std::vector<actor>{caf::actor_cast<caf::actor>(tracker)};
+    if (auto acc = self->system().registry().get(accountant_atom::value))
```
[the remark this answers is not in the document]
Comment (mavam) [reply libvast/src/system/node.cpp:390]: I couldn't see that. The tracker doesn't have the accountant on file from what I can tell.

File: libvast/src/system/terminator.cpp [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+    if (remaining.empty()) {
+      VAST_DEBUG(self, "quits prematurely because all actors have exited");
+      self->state.promise.deliver(done_atom::value);
+      self->quit(caf::exit_reason::user_shutdown);
+      return;
+    }
+    if constexpr (std::is_same_v<Policy, policy::sequential>) {
+      // Track actors in reverse order because the user provides the actors in
+      // the order of shutdown, but we use a stack internally that stores the
+      // first actor to be terminated at the end.
```
[the remark this answers is not in the document]
Comment (mavam) [reply libvast/src/system/terminator.cpp:82]: The reason here is more convenience: since a `vector<T>` supports efficient push/pop at the end, I decided to put the actors in reverse order so that killing the next actor results in popping from the stack.

This argument doesn't hold for parallel shutdown, though.

File: libvast/src/system/node.cpp [hunk 5]
```diff
@@ -389,13 +383,23 @@ void node_state::init(std::string init_name, path init_dir) {
...
   // Bring up the tracker.
-  tracker = self->spawn<monitored>(system::tracker, name);
-  self->set_down_handler([=](const down_msg& msg) {
-    VAST_IGNORE_UNUSED(msg);
-    VAST_DEBUG(self, "got DOWN from", msg.source);
-    self->quit(msg.reason);
+  tracker = self->spawn<linked>(system::tracker, name);
+  self->set_exit_handler([=](const exit_msg& msg) {
+    VAST_DEBUG(self, "got EXIT from", msg.source);
+    auto xs = std::vector<actor>{caf::actor_cast<caf::actor>(tracker)};
```
[the remark this answers is not in the document]
Comment (mavam) [reply libvast/src/system/node.cpp]: Yikes :-)

File: libvast/src/system/tracker.cpp [hunk 6]
```diff
@@ -188,20 +102,30 @@ tracker(tracker_type::stateful_pointer<tracker_state> self, std::string node) {
...
-                  self->state.registry.components[node],
-                  std::vector<std::string>{"exporter", "index", "archive",
-                                           "importer", "source"});
+  self->set_exit_handler([=](const exit_msg&) {
+    // Only trap the first exit, regularly shutdown on the next one.
+    self->set_exit_handler({});
+    // We shut down the components in the order in which data flows so that
+    // downstream components can still process in-flight data. Because the
+    // terminator operates with a stack of components, we specify them in
+    // reverse order.
```
[the remark this answers is not in the document]
Comment (mavam) [reply libvast/src/system/tracker.cpp]: Good catch; removed.

File: libvast/src/system/terminator.cpp [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+    }
+    if (remaining.empty()) {
+      VAST_DEBUG(self, "terminated all actors");
+      self->state.promise.deliver(done_atom::value);
+      self->quit(caf::exit_reason::user_shutdown);
+    }
+  });
+  return {[=](const std::vector<caf::actor>& xs) {
+    VAST_DEBUG(self, "got request to terminate", xs.size(), "actors");
+    self->state.promise = self->make_response_promise();
```
[the remark this answers is not in the document]
Comment (mavam) [reply libvast/src/system/terminator.cpp:63]: Technically, yes. But spawning a terminator is usually a "throwaway" actor that you interact exactly once.

I've added an assertion to make sure this usage error gets caught early.

[the remark this answers is not in the document]
Comment (dominiklohmann) [reply libvast/src/system/terminator.cpp:63]: Chiming in here: I don't think we should add assertions in the shutdown path, as that can lead to data loss.

[the remark this answers is not in the document]
Comment (dominiklohmann) [reply libvast/src/system/terminator.cpp:63]: Also, shouldn't this check be running in release mode as well?

File: libvast/src/system/node.cpp [hunk 8]
```diff
@@ -389,13 +383,23 @@ void node_state::init(std::string init_name, path init_dir) {
...
+    if (auto acc = self->system().registry().get(accountant_atom::value))
+      xs.push_back(caf::actor_cast<caf::actor>(acc));
+    auto t = self->spawn(terminator<policy::sequential>);
+    shutdown(self, t, std::move(xs));
+    self->attach_functor([=](const caf::error& reason) {
+      VAST_DEBUG(self, "terminated all dependent actors");
+      self->quit(reason);
```
[the remark this answers is not in the document]
Comment (mavam) [reply libvast/src/system/node.cpp]: Yep, fixed.

I don't quite understand why we would want to add `self` to the list of actors we are shutting down. To me, it feels a more natural that we wait with the terminator until all children have terminated, and then exit in the parent context via `self->quit`.

[the remark this answers is not in the document]
Comment (mavam) [reply libvast/src/system/node.cpp]: Using `attach_functor` here is just a workaround to fix the actor system registry bug. Once it's fixed, we can get rid of it entirely.

File: libvast/src/system/node.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (lava) [reply libvast/src/system/node.cpp:390]: That's odd, I had this removed locally and the accountant was getting killed. Maybe there are some other factors that influence this.

File: libvast/src/system/node.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (lava) [reply libvast/src/system/node.cpp:200]: I agree that it shouldnt matter that we get multiple EXIT messages, but in this case linking also implies that if the accountant goes down, the node will go down. Do we want that behaviour?

**Comment by dominiklohmann** [comment]
Note that some changes had to be made for the `component_map*` from `tracker.hpp` for #901 (see commit https://github.com/tenzir/vast/pull/901/commits/3f6edb04d9f541901df3b91da97d826bbdd07333).

File: libvast/vast/system/terminator.hpp (hunk 1 above)
Comment (mavam) [reply libvast/vast/system/terminator.hpp]: I've modularized the terminator actor a bit. There are now different helper utilities to interact with it:

- `shutdown`: non-blocking termination that also terminates the caller.
- `wait`: blocking termination modeled after `scoped_actor::wait_for`.

All the helper functions include spawning the `terminator`, so it's no longer necessary at the call site. Looks way more reasonable that way.

File: libvast/src/system/terminator.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (mavam) [reply libvast/src/system/terminator.cpp:66]: >  I don't think we should add assertions in the shutdown path, as that can lead to data loss.

I think an assertion is the correct mechanism here. If a developer uses the API of the terminator incorrectly, all bets are off. Then you should terminate immediately.

File: libvast/src/system/tracker.cpp [hunk 9]
```diff
@@ -188,20 +102,30 @@ tracker(tracker_type::stateful_pointer<tracker_state> self, std::string node) {
...
+    auto actors = std::vector<caf::actor>{};
+    auto components = std::vector<std::string>{"source", "importer", "archive",
+                                               "index", "exporter"};
```
[the remark this answers is not in the document]
Comment (mavam) [reply libvast/src/system/tracker.cpp:110]: In my mental model, the accountant is like a source. We currently terminate it before the tracker goes down. Does that work in your opinion as well?

File: libvast/src/system/terminator.cpp [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+    } else {
+      static_assert(detail::always_false_v<Policy>, "unsupported policy");
+    }
+    if (remaining.empty()) {
+      VAST_DEBUG(self, "terminated all actors");
+      self->state.promise.deliver(done_atom::value);
+      self->quit(caf::exit_reason::user_shutdown);
+    }
+  });
+  return {[=](const std::vector<caf::actor>& xs) {
```
[the remark this answers is not in the document]
Comment (mavam) [reply libvast/src/system/terminator.cpp:63]: We still need to send the list of actors to the terminator. This means you have to make at least one copy by encapsulating the vector into a message. We capture the message contents as const ref here, which is a separate path when the runtime kicks in, independent of the vector that is passed initially.

**Comment by mavam** [comment]
@dominiklohmann thanks for pointing this out. Could ask I ask you to add a commit to this branch that makes this PR mergeable?

**Comment by dominiklohmann** [comment]
Sure thing. Once you're done, hand me the edit token and I'll do the rebase. The atom usage needs to be changed now, too.

**Comment by mavam** [comment]
You got the token!

**Comment by dominiklohmann** [comment]
I'm done.

File: libvast/src/system/node.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (mavam) [reply libvast/src/system/node.cpp:195]: I think we do. The accountant is a low-complexity piece and should be robust. If it fails, we have a bigger problem that we probably shouldn't ignore. Therefore, I suggest fate sharing.

File: libvast/src/system/node.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (mavam) [reply libvast/src/system/node.cpp]: Maybe because the accountant is linked to the node?

File: libvast/src/system/tracker.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (lava) [reply libvast/src/system/tracker.cpp:110]: Thats fine then, but maybe warrants a comment in `node.cpp` that the order is deliberate and important.

File: libvast/src/system/node.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (mavam) [reply libvast/src/system/node.cpp]: Seems like this is solved.

**Comment by mavam** [comment]
@dominiklohmann we talked about the actor typing today, and we want do it within this PR.

**Comment by mavam** [comment]
@lava I think I've addressed all you changes. `wait` is now called `terminate`. Only CI now needs to give green light. Then we're done.

File: libvast/src/system/node.cpp [hunk 11]
```diff
@@ -381,13 +375,22 @@ void node_state::init(std::string init_name, path init_dir) {
...
-  self->set_down_handler([=](const down_msg& msg) {
-    VAST_IGNORE_UNUSED(msg);
-    VAST_DEBUG(self, "got DOWN from", msg.source);
-    self->quit(msg.reason);
+  tracker = self->spawn<linked>(system::tracker, name);
+  self->set_exit_handler([=](const exit_msg& msg) {
+    VAST_DEBUG(self, "got EXIT from", msg.source);
+    std::vector<caf::actor> actors;
+    if (auto acc = self->system().registry().get(atom::accountant_v))
+      actors.push_back(caf::actor_cast<caf::actor>(acc));
```
Comment (lava) [inline libvast/src/system/node.cpp:383]: We could mention in a comment here that we intentionally use a sequential shutdown policy and put the accountant first to allow it to clear its buffers before the index, archive, etc. shut down.

File: libvast/src/system/profiler.cpp [hunk 12]
```diff
@@ -20,6 +20,7 @@
 #include "vast/logger.hpp"
 #include "vast/time.hpp"
 
+#include <caf/event_based_actor.hpp>
```
Comment (lava) [inline libvast/src/system/profiler.cpp:23]: Unnecessary include.

File: libvast/src/system/terminator.cpp [hunk 13]
```diff
@@ -0,0 +1,107 @@
...
+#include "vast/fwd.hpp"
+#include "vast/logger.hpp"
+
+#include <caf/behavior.hpp>
+#include <caf/event_based_actor.hpp>
+#include <caf/response_promise.hpp>
+
+namespace vast::system {
+
+/// Performs a parallel shutdown of a list of actors.
```
Comment (lava) [inline libvast/src/system/terminator.cpp:26]: A "parallel" shutdown is not always true.

File: libvast/vast/system/shutdown.hpp [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+namespace vast::policy {
+
+struct sequential;
+struct parallel;
+
+} // namespace vast::policy
+
+namespace vast::system {
+
+/// Performs an asynchronous shutdown of an actor by first shutting down a set
```
Comment (lava) [inline libvast/vast/system/shutdown.hpp:32]: Not a native speaker, but `by` sounds like the wrong preposition.

File: libvast/vast/system/shutdown.hpp [hunk 15]
```diff
@@ -0,0 +1,48 @@
...
+/// of dependent actors. The shutdown process runs either sequentially or in
+/// parallel. As soon as all dependent actors have terminated, the calling
+/// actor exits.
+/// @param self The actor to terminate.
+/// @param xs Owned actors by *self* that need to shutdown prior to *self*.
+template <class Policy>
+void shutdown(caf::event_based_actor* self, std::vector<caf::actor> xs);
```
Comment (lava) [inline libvast/vast/system/shutdown.hpp:39]: After thinking about it, I'm not sure this works as you intended: You can call shutdown with any `Policy`, but unless it is `sequential` or `parallel` it should generate a linker error because these are the only ones generated in `shutdown.cpp`.

Probably the correct way would be to say
```
template<class Policy>
void shutdown(...) = delete;

template<>
void shutdown<policy::sequential>(...);


template<>
void shutdown<policy::parallel>(...);
```

(although this is only theorycrafting, I didn't test it with another type)

File: libvast/vast/system/terminate.hpp [hunk 16]
```diff
@@ -0,0 +1,68 @@
...
+
+} // namespace vast::policy
+
+namespace vast::system {
+
+/// Performs an asynchronous shutdown of a set of dependent actors,
+/// configurable either in sequential or parallel mode of operation. As soon as
+/// all dependent actors have terminated, the returned promise gets fulfilled.
+/// @param self The actor to terminate.
+/// @param xs Owned actors by *self* that need to shutdown prior to *self*.
```
Comment (lava) [inline libvast/vast/system/terminate.hpp:42]: The "owning" part seems to be unrelated to this function; here `*self* ` is just an arbitrary actor that is used to dispatch messages. (i.e. one might just as well use this function with a scoped_actor to kill a bunch of other actors in sequence)

File: libvast/vast/system/terminate.hpp [hunk 17]
```diff
@@ -0,0 +1,68 @@
...
+struct sequential;
+struct parallel;
+
+} // namespace vast::policy
+
+namespace vast::system {
+
+/// Performs an asynchronous shutdown of a set of dependent actors,
+/// configurable either in sequential or parallel mode of operation. As soon as
+/// all dependent actors have terminated, the returned promise gets fulfilled.
```
Comment (lava) [inline libvast/vast/system/terminate.hpp:40]: I think we can just remove the word `dependent` here. (also we could mention that its going to perform the shutdown by sending `exit` messages and aborting unsuccessfully after a timeout was reached)

File: libvast/vast/system/shutdown.hpp (hunk 15 above)
Comment (mavam) [reply libvast/vast/system/shutdown.hpp:39]: There's a `static_assert` that fires if the policy is neither `sequential` nor `parallel`.

File: libvast/src/system/profiler.cpp (hunk 12 above)
Comment (mavam) [reply libvast/src/system/profiler.cpp:23]: Actually, this is now necessary; see #928 from where I cherry-picked this commit for this branch to compile.

File: libvast/vast/system/shutdown.hpp (hunk 15 above)
Comment (lava) [reply libvast/vast/system/shutdown.hpp:39]: I don't see how that can work, the `static_assert` is in a different translation unit so it should not be visible at the call site of, e.g.,  `shutdown<int>()`.

Comment (dominiklohmann) [reply libvast/vast/system/shutdown.hpp:39]: This would be the correct way to express the `= delete` here if I recall correctly (not tested):

```cpp
template <class Policy>
void shutdown(...) {
  static_assert(detail::always_false_v<Policy>, "unsupported policy");
}
```

**Comment by mavam** [comment]
I've addressed your comments, @lava. Small but important points that help users understand the API correctly.
